library/core/src/num/int_macros.rs

```diff
@@ -2766,14 +2777,21 @@ macro_rules! int_impl {
                       without modifying the original"]
         #[inline]
         pub const fn isqrt(self) -> Self {
-            // I would like to implement it as
-            // ```
-            // self.checked_isqrt().expect("argument of integer square root must be non-negative")
-            // ```
-            // but `expect` is not yet stable as a `const fn`.
-            match self.checked_isqrt() {
-                Some(sqrt) => sqrt,
-                None => panic!("argument of integer square root must be non-negative"),
+            if self < 0 {
+                crate::num::int_sqrt::panic_for_negative_argument();
+            } else {
+                let result = crate::num::int_sqrt::$ActualT(self as $ActualT) as $SelfT;
+
+                // SAFETY: Inform the optimizer that square roots of
+                // nonnegative integers are nonnegative and what the maximum
+                // result is.
+                unsafe {
+                    crate::hint::assert_unchecked(result >= 0);
+                    const MAX_RESULT: $SelfT = crate::num::int_sqrt::$ActualT($ActualT::MAX) as $SelfT;
+                    crate::hint::assert_unchecked(result <= MAX_RESULT);
+                }
+
+                result
```

Comment:
Why can't this just be left as-is, in terms of `checked_isqrt`?

Reply:
The main reason is so that I can control inlining.

I put `#[inline(always)]` above all the non-"intermediate" functions in `library/core/src/num/int_sqrt.rs` so that those functions would be inlined into the `checked_isqrt` and `isqrt` functions, as if they'd been custom written and to reduce the number of function calls.

To do that in `isqrt` indirectly through `checked_isqrt`, I'd need to mark `checked_isqrt` as `#[inline(always)]`, but that would cause the behind-the-scenes implementations to be inlined wherever the programmer used `checked_isqrt`.

Along those lines, I still need to remove the `#[inline]` attributes on `isqrt` and `checked_isqrt`.

---

The "intermediate" functions support the types with more bits. For example, `u128` uses `intermediate_u64`, which uses `intermediate_u32`, which uses `intermediate_u16`, which uses `intermediate_u8`. The intermediate functions produce the integer square root and the remainder (how much higher the input is than the square of its integer square root).

I don't want to inline the intermediate functions because then `u128` would become huge. Not inlining them also allows them to be shared if, for example, the programmer uses `isqrt` on both `u16`s and `u64`s.

---

Do you think that this is a good plan?

Reply:
I figured it was something inline-related. Could you add nonpublic a `isqrt_inner` to `NonZero*` that is `#[inline(always)]`, then call that from `NonZero::<T>::isqrt`, `T::checked_isqrt`, and `T::isqrt`? So at least the logic stays in only one place.

It is probably okay to leave the public functions `#[inline]` since that just allows cross-crate inlining, doesn't mean it has to be. And for the intermediate functions, it is probably okay to just not add any `inline` and let LLVM pick whether to inline them or not. Unless checking the binary size shows it makes bad decisions here.

Reply:
I've tried out your suggestion to restore `i*::isqrt`'s call to `checked_isqrt`, and it looks like it still inlines as I want it to (at least when I tested `i*::sqrt` and `i*::checked_isqrt`), so I've put it back that way.

I've manually inlined the `intermediate` functions into the unsigned functions and then removed them for an overall speedup (except a slight slowdown for small inputs on some types). Doing so eliminates some repetitive work, as we can normalize (by shifting left) and denormalize (by shifting right) only once for the whole process.

As far as having it in one place, I think that leaving it in `library/core/src/num/int_sqrt.rs` and calling it from there is a good idea now, as the inlining happens nicely with it set up that way.

Reply:
OK, I've rechecked and saw that I didn't really restore the `isqrt` method to the prior form. I've now restored it to the old `match`, but with the `#[cold]` and `#[track_caller]` panic function for negative inputs. I got that idea from the `checked_ilog*` methods.

Reply:
Awesome! One followup in a separate comment, I'll mark this one resolved.